### PR TITLE
Add support for multi-value header routing

### LINKED ISSUE #1494: Multi-value header routing support
### Describe the bug
We've been looking at a very weird issue in YARP's routing almost the entire day and we still haven't quite pinpointed yet what the problem is, but we can consistently reproduce the issue in our test environments.

The problem stems from the fact that we have two routes, which are mostly identical. Here is route1:

```json
{
   "RouteId": "route1",
   "Match": {
     "Path": "/api/{**catch-all}",             
   },
   "ClusterId": "api",
   "CorsPolicy": "default",
   "Transforms": [
     {
       "PathRemovePrefix": "/api"
     },
     {
       "X-Forwarded": "Off"
     }
   ]
}
```

And route 2:

```json
{
  "RouteId": "route2",
  "Match": {
    "Path": "/api/{**catch-all}",
    "Headers": [
       {
         "Name": "Cookie",
         "Values": [
           ".AspNetCore.Cookies"
         ],
         "Mode": "Contains"
        }
      ]
    },
    "ClusterId": "api",
    "AuthorizationPolicy": "CookieAuthentication",
    "CorsPolicy": "default",
    "Transforms": [
       {
         "PathRemovePrefix": "/api"
       },
       {
         "X-Forwarded": "Off"
       }
    ]
}
```

As you can see both routes are nearly identical, but `route2` has an additional match based on Headers. This has been working fine for us so far while running on Windows and mostly using HTTP/1.1. We are now trying to migrate this to run on Linux containers on Kubernetes with HTTP/2 and that's where we are seeing something interesting. In particular `route2` never seems to be chosen at least not while sending requests from Chrome and Edge. Sending a request with cURL using the exact same headers though seems to work just fine, which is what makes this such a weird issue.

For example, here is the output of the HttpRequestLogging middleware for a request done using Chrome:

```json
{
    "@t": "2022-01-05T14:12:16.3714830Z",
    "@m": "Request:\nProtocol: HTTP/2\nMethod: GET\nScheme: https\nPathBase: \nPath: /api/user\nAccept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9\nHost: services.apollo.zeus.corp\nUser-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/96.0.4664.110 Safari/537.36\n:method: [Redacted]\nAccept-Encoding: gzip, deflate, br\nAccept-Language: nl-NL,nl;q=0.9,en-US;q=0.8,en;q=0.7\nCache-Control: max-age=0\nCookie: .AspNetCore.Cookies=chunks-2,.AspNetCore.CookiesC1=<sensitive-value>,.AspNetCore.CookiesC2=<sensitive-value>\nUpgrade-Insecure-Requests: [Redacted]\nsec-ch-ua: [Redacted]\nsec-ch-ua-mobile: [Redacted]\nsec-ch-ua-platform: [Redacted]\nsec-fetch-site: [Redacted]\nsec-fetch-mode: [Redacted]\nsec-fetch-user: [Redacted]\nsec-fetch-dest: [Redacted]",
    "@i": "893de8da",
    "Protocol": "HTTP/2",
    "Method": "GET",
    "Scheme": "https",
    "PathBase": "",
    "Path": "/api/user",
    "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9",
    "Host": "services.apollo.zeus.corp",
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/96.0.4664.110 Safari/537.36",
    ":method": "[Redacted]",
    "Accept-Encoding": "gzip, deflate, br",
    "Accept-Language": "nl-NL,nl;q=0.9,en-US;q=0.8,en;q=0.7",
    "Cache-Control": "max-age=0",
    "Cookie": ".AspNetCore.Cookies=chunks-2,.AspNetCore.CookiesC1=<sensitive-value>,.AspNetCore.CookiesC2=<sensitive-value>",
    "Upgrade-Insecure-Requests": "[Redacted]",
    "sec-ch-ua": "[Redacted]",
    "sec-ch-ua-mobile": "[Redacted]",
    "sec-ch-ua-platform": "[Redacted]",
    "sec-fetch-site": "[Redacted]",
    "sec-fetch-mode": "[Redacted]",
    "sec-fetch-user": "[Redacted]",
    "sec-fetch-dest": "[Redacted]",
    "HttpRequestLog": "Request:\nProtocol: HTTP/2\nMethod: GET\nScheme: https\nPathBase: \nPath: /api/user\nAccept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9\nHost: services.apollo.zeus.corp\nUser-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/96.0.4664.110 Safari/537.36\n:method: [Redacted]\nAccept-Encoding: gzip, deflate, br\nAccept-Language: nl-NL,nl;q=0.9,en-US;q=0.8,en;q=0.7\nCache-Control: max-age=0\nCookie: .AspNetCore.Cookies=chunks-2,.AspNetCore.CookiesC1=<sensitive-value>,.AspNetCore.CookiesC2=<sensitive-value>\nUpgrade-Insecure-Requests: [Redacted]\nsec-ch-ua: [Redacted]\nsec-ch-ua-mobile: [Redacted]\nsec-ch-ua-platform: [Redacted]\nsec-fetch-site: [Redacted]\nsec-fetch-mode: [Redacted]\nsec-fetch-user: [Redacted]\nsec-fetch-dest: [Redacted]",
    "EventId": {
        "Id": 1,
        "Name": "RequestLog"
    },
    "SourceContext": "Microsoft.AspNetCore.HttpLogging.HttpLoggingMiddleware",
    "RequestId": "0HMEGAFLMA6A6:00000001",
    "RequestPath": "/api/user",
    "ConnectionId": "0HMEGAFLMA6A6",
    "MachineName": "apigateway-f8dcddfd6-lc2gn",
    "ProcessId": 1,
    "ThreadId": 33,
    "LogicalServiceName": "API Gateway"
}
```

Here is the same request done using cURL (by copying the request from Chrome Developer Tools as a cURL command):

```json
{
    "@t": "2022-01-05T14:10:13.6241996Z",
    "@m": "Request:\nProtocol: HTTP/2\nMethod: GET\nScheme: https\nPathBase: \nPath: /api/user\nAccept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9\nHost: services.apollo.zeus.corp\nUser-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/96.0.4664.110 Safari/537.36\n:method: [Redacted]\nAccept-Encoding: deflate, gzip, br, zstd\nAccept-Language: nl-NL,nl;q=0.9,en-US;q=0.8,en;q=0.7\nCache-Control: max-age=0\nCookie: .AspNetCore.Cookies=chunks-2,.AspNetCore.CookiesC1=<sensitive-value>,.AspNetCore.CookiesC2=<sensitive-value>\nUpgrade-Insecure-Requests: [Redacted]\nauthority: [Redacted]\nsec-ch-ua-mobile: [Redacted]\nsec-ch-ua-platform: [Redacted]\nsec-fetch-site: [Redacted]\nsec-fetch-mode: [Redacted]\nsec-fetch-user: [Redacted]\nsec-fetch-dest: [Redacted]",
    "@i": "893de8da",
    "Protocol": "HTTP/2",
    "Method": "GET",
    "Scheme": "https",
    "PathBase": "",
    "Path": "/api/user",
    "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9",
    "Host": "services.apollo.zeus.corp",
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/96.0.4664.110 Safari/537.36",
    ":method": "[Redacted]",
    "Accept-Encoding": "deflate, gzip, br, zstd",
    "Accept-Language": "nl-NL,nl;q=0.9,en-US;q=0.8,en;q=0.7",
    "Cache-Control": "max-age=0",
    "Cookie": ".AspNetCore.Cookies=chunks-2,.AspNetCore.CookiesC1=<sensitive-value>,.AspNetCore.CookiesC2=<sensitive-value>",
    "Upgrade-Insecure-Requests": "[Redacted]",
    "authority": "[Redacted]",
    "sec-ch-ua-mobile": "[Redacted]",
    "sec-ch-ua-platform": "[Redacted]",
    "sec-fetch-site": "[Redacted]",
    "sec-fetch-mode": "[Redacted]",
    "sec-fetch-user": "[Redacted]",
    "sec-fetch-dest": "[Redacted]",
    "HttpRequestLog": "Request:\nProtocol: HTTP/2\nMethod: GET\nScheme: https\nPathBase: \nPath: /api/user\nAccept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9\nHost: services.apollo.zeus.corp\nUser-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/96.0.4664.110 Safari/537.36\n:method: [Redacted]\nAccept-Encoding: deflate, gzip, br, zstd\nAccept-Language: nl-NL,nl;q=0.9,en-US;q=0.8,en;q=0.7\nCache-Control: max-age=0\nCookie: .AspNetCore.Cookies=chunks-2,.AspNetCore.CookiesC1=<sensitive-value>,.AspNetCore.CookiesC2=<sensitive-value>\nUpgrade-Insecure-Requests: [Redacted]\nauthority: [Redacted]\nsec-ch-ua-mobile: [Redacted]\nsec-ch-ua-platform: [Redacted]\nsec-fetch-site: [Redacted]\nsec-fetch-mode: [Redacted]\nsec-fetch-user: [Redacted]\nsec-fetch-dest: [Redacted]",
    "EventId": {
        "Id": 1,
        "Name": "RequestLog"
    },
    "SourceContext": "Microsoft.AspNetCore.HttpLogging.HttpLoggingMiddleware",
    "RequestId": "0HMEGAGUUK0CE:00000001",
    "RequestPath": "/api/user",
    "ConnectionId": "0HMEGAGUUK0CE",
    "MachineName": "apigateway-f8dcddfd6-bfznx",
    "ProcessId": 1,
    "ThreadId": 41,
    "LogicalServiceName": "API Gateway"
}
```

For that first request `route1` is chosen as the endpoint, while for the second request `route2` is chosen. So far we haven't been able to figure out what makes these requests different.

### To Reproduce
We've tried to create a small repro, but so far haven't been able to. Even with a minimal .NET 6 project with YARP added to it and putting that into a Docker container and doing a request with cURL works just fine.

### Further technical details

- Include the version of the packages you are using: 1.0.0 on .NET 6.0.1
- The platform (Linux/macOS/Windows): Linux (containers, alpine)

### PR DESCRIPTION
**Pull request by MihaZupan**
Fixes #1494

Matching now treats multiple values as *OR*s. I.e. `Exact`/`Prefix` must match *any* of the request values.
`NotContains` means the value must *not match any* request header.

Values are further split based on the separator `,` (or `;` for `Cookie`). A single outer pair of quotes is also removed before matching.

This PR is generally slightly faster on simple (single-value) inputs, and (as expected) slower on multi-value ones.
Modes that may observe a perf degradation are `Exact` and `Prefix`.
I don't think the difference is prohibitive to the point that having an opt-out switch is needed.

Benchmark scenarios are defined [here](https://gist.github.com/MihaZupan/ccac2200ff7ba2f509060ee1a061e230#file-headerroutingbenchmark-cs-L30-L64).

| Method |         Scenario |      Mean |    Error |   StdDev | Ratio |
|------- |----------------- |----------:|---------:|---------:|------:|
|    Old |   SimpleContains | 292.78 ns | 0.957 ns | 0.848 ns |  1.00 |
|    New |   SimpleContains | 243.82 ns | 0.823 ns | 0.687 ns |  0.83 |
|        |                  |           |          |          |       |
|    Old |      SimpleExact | 158.98 ns | 0.529 ns | 0.469 ns |  1.00 |
|    New |      SimpleExact | 117.40 ns | 0.236 ns | 0.184 ns |  0.74 |
|        |                  |           |          |          |       |
|    Old |     SimplePrefix | 104.09 ns | 0.284 ns | 0.222 ns |  1.00 |
|    New |     SimplePrefix |  97.69 ns | 0.512 ns | 0.454 ns |  0.94 |
|        |                  |           |          |          |       |
|    Old | MultipleContains | 294.57 ns | 1.679 ns | 1.402 ns |  1.00 |
|    New | MultipleContains | 291.51 ns | 3.079 ns | 2.880 ns |  0.99 |
|        |                  |           |          |          |       |
|    Old |    MultipleExact | 124.28 ns | 0.237 ns | 0.185 ns |  1.00 |
|    New |    MultipleExact | 175.34 ns | 0.626 ns | 0.523 ns |  1.41 |
|        |                  |           |          |          |       |
|    Old |   MultiplePrefix | 130.89 ns | 1.529 ns | 1.431 ns |  1.00 |
|    New |   MultiplePrefix | 176.16 ns | 0.513 ns | 0.428 ns |  1.35 |

ToDo: Updating docs with the new behavior

### CODE REVIEW

File: src/ReverseProxy/Routing/HeaderMatcher.cs [hunk 1]
```diff
@@ -30,11 +32,18 @@ public HeaderMatcher(string name, IReadOnlyList<string>? values, HeaderMatchMode
...
+        }
 
         Name = name;
-        Values = values ?? Array.Empty<string>();
+        Values = values?.ToArray() ?? Array.Empty<string>();
         Mode = mode;
-        IsCaseSensitive = isCaseSensitive;
+        Comparison = isCaseSensitive ? StringComparison.Ordinal : StringComparison.OrdinalIgnoreCase;
+
+        // Should this extend to Set-Cookie?
```
Comment (Tratcher) [inline src/ReverseProxy/Routing/HeaderMatcher.cs:45]: No, Set-Cookie is a response header, it shouldn't show up in routing. It has a different format regardless.
```suggestion
```

File: src/ReverseProxy/Routing/HeaderMatcherPolicy.cs [hunk 2]
```diff
@@ -51,6 +52,13 @@ public Task ApplyAsync(HttpContext httpContext, CandidateSet candidates)
         _ = httpContext ?? throw new ArgumentNullException(nameof(httpContext));
         _ = candidates ?? throw new ArgumentNullException(nameof(candidates));
 
+        ApplyCore(candidates, httpContext.Request.Headers);
+
+        return Task.CompletedTask;
+    }
+
+    private static void ApplyCore(CandidateSet candidates, IHeaderDictionary headers)
+    {
```
Comment (Tratcher) [inline src/ReverseProxy/Routing/HeaderMatcherPolicy.cs:61]: Why is ApplyCore a separate method?

**Review by Tratcher — APPROVED** [review]

File: src/ReverseProxy/Routing/HeaderMatcherPolicy.cs (hunk 2 above)
Comment (MihaZupan) [reply src/ReverseProxy/Routing/HeaderMatcherPolicy.cs:61]: I believe it's just slightly faster - not a huge deal either way. Want me to revert it?

Comment (Tratcher) [reply src/ReverseProxy/Routing/HeaderMatcherPolicy.cs:61]: Yes, please collapse these two methods. The benefits are insubstantial and it's unnecessary.

**Pull request merged**